metrics/internal/stackdriver/secret_watcher_test.go

```diff
@@ -0,0 +1,569 @@
+	}
+
+	defaultWatchersToTest = []struct {
+		name        string
+		constructor func(t *testing.T, obs ...Observer) SecretWatcher
+	}{
+		{
+			name:        "GlobalSecretWatcher",
+			constructor: defaultGlobalSecretWatcherConstructor,
+		}, {
+			name:        "SingleNamespaceSecretWatcher",
+			constructor: defaultSingleSecretSecretWatcherConstructor,
+		}, {
+			name:        "SingleSecretSecretWatcher",
+			constructor: defaultSingleSecretSecretWatcherConstructor,
+		},
+	}
+
+	defaultObserverFuncs = &ObserverFuncs{
+		AddFunc: func(s *corev1.Secret) {
+		},
+		UpdateFunc: func(sOld *corev1.Secret, sNew *corev1.Secret) {
+		},
+		DeleteFunc: func(s *corev1.Secret) {
+		},
+	}
+)
+
+// secretForTest encapsulates Secret metadata and a Secret for re-using and verifying values.
+type secretForTest struct {
+	namespace string
+	name      string
+	dataKey   string
+	dataValue string
+	secret    corev1.Secret
+}
+
+func (s *secretForTest) UpdateDataKey(dataKey string) *secretForTest {
+	s.dataKey = dataKey
+	s.secret.Data = map[string][]byte{
+		s.dataKey: []byte(s.dataValue),
+	}
+
+	return s
+}
+
+func (s *secretForTest) UpdateDataValue(dataValue string) *secretForTest {
+	s.dataValue = dataValue
+	s.secret.Data = map[string][]byte{
+		s.dataKey: []byte(s.dataValue),
+	}
+
+	return s
+}
+
+// newSecretForTest constructs a secretForTest.
+func newSecretForTest(namespace string, name string) *secretForTest {
+	return &secretForTest{
+		namespace: namespace,
+		name:      name,
+		dataKey:   defaultSecretDataKey,
+		dataValue: defaultSecretDataValue,
+		secret: corev1.Secret{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Secret",
+				APIVersion: "apps/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+			Data: map[string][]byte{
+				defaultSecretDataKey: []byte(defaultSecretDataValue),
+			},
+			Type: "Opaque",
+		},
+	}
+}
+
+// testObserver is an implementation of the the secretWatcher observer interface
+// that also holds state about whether the observer's callbacks have been triggered.
+type testObserver struct {
+	oFuncs         *ObserverFuncs
+	onAddCalled    bool
+	onUpdateCalled bool
+	onDeleteCalled bool
+	mux            sync.Mutex
+}
+
+// newTestObserver constructs a testObserver from observerFuncs.
+func newTestObserver(inputObsFuncs *ObserverFuncs) *testObserver {
+	// If a callback is nil, consider it to be called by default.
+	return &testObserver{
+		oFuncs:         inputObsFuncs,
+		onAddCalled:    inputObsFuncs.AddFunc == nil,
+		onUpdateCalled: inputObsFuncs.UpdateFunc == nil,
+		onDeleteCalled: inputObsFuncs.DeleteFunc == nil,
+	}
+}
+
+func (tObs *testObserver) markCallbackCalled(callbackSentinel *bool) {
+	tObs.mux.Lock()
+	defer tObs.mux.Unlock()
+	*callbackSentinel = true
+}
+
+func (tObs *testObserver) OnAdd(s *corev1.Secret) {
+	tObs.oFuncs.OnAdd(s)
+	tObs.markCallbackCalled(&tObs.onAddCalled)
```

Comment:
I prefer to just use `tObs.onAddCalled = true` and add the lock around and get rid of  `markCallbackCalled`.

Reply:
markCallbackCalled exists to prevent code duplication of the lock/unlock call. Otherwise each of OnAdd, OnUpdate, and OnDelete would have to reimplement that part. (Without the lock, gotest will report a Data Race when trying to mark the callback as called).